### PR TITLE
playwright setup and implementation of listFolder Step

### PR DESCRIPTION
**Pull request by S-Panta**
This PR adds a UI test for moodle integration with Ocis regarding Folder listing.

### CODE REVIEW

File: cucumber.conf.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    global.browser = await chromium.launch({
+        headless: false,
+        slowMo: 1000,
+    });
+});
+
+AfterAll(async function () {
+    await global.browser.close();
+});
+Before(async function () {
```
Comment (saw-jan) [inline cucumber.conf.js:15]:
```suggestion
});

Before(async function () {
```

File: cucumber.conf.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+const { Before, BeforeAll, AfterAll, After, setDefaultTimeout } = require("@cucumber/cucumber");
+const { chromium } = require("playwright");
+
+setDefaultTimeout(60000)
+BeforeAll(async function () {
```
Comment (saw-jan) [inline cucumber.conf.js:5]:
```suggestion
setDefaultTimeout(60000)

BeforeAll(async function () {
```

File: tests/cucumber/acceptance/stepDefinitions/listfolder.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    await login.loginUser();
+    await expect(page).toHaveURL(login.homePageUrl);
+});
+
+Given('the user has navigated to add a new course page', async function () {
+    await myCourse.navigateToMyCourseMenu();
+    await expect(page).toHaveURL(myCourse.myCourseUrl);
+});
+
+When('the user clicks file-picker', async function () {
```
Comment (individual-it) [inline tests/cucumber/acceptance/stepDefinitions/listfolder.js:24]:
```suggestion
When('the user opens the file-picker', async function () {
```

File: tests/cucumber/acceptance/pageObjects/FilePicker.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+class FilePicker {
+    constructor() {
+        this.fileUploadSelector = 'span.fp-repo-name:has-text("owncloud")';
+        this.fileListSelector ='//div[@class="fp-tableview"]//span[@class="fp-filename"]';
+        this.viewByListFileSelector = 'a[role="button"][title="Display folder with file details"]'
+    }
+
+    async selectRepositoryForUpload(){
```
Comment (individual-it) [inline tests/cucumber/acceptance/pageObjects/FilePicker.js:8]: its not for upload

```suggestion
    async selectRepository(){
```

also make the name dynamic

File: tests/cucumber/acceptance/stepDefinitions/listfolder.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    await myCourse.navigateToFilePicker();
+});
+
+When('the user selects Owncloud from the sidebar menu', async function () {
```
Comment (individual-it) [inline tests/cucumber/acceptance/stepDefinitions/listfolder.js:28]: make the repository name dynamig
```suggestion
When('the user selects the repository "owncloud"', async function (repository) {
```

File: tests/cucumber/acceptance/stepDefinitions/listfolder.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    await filePicker.selectRepositoryForUpload();
+});
+
+Then('the file list from owncloud should be seen on the webUI', async function () {
```
Comment (individual-it) [inline tests/cucumber/acceptance/stepDefinitions/listfolder.js:32]:
```suggestion
Then('the file list should be displayed', async function () {
```

**Review by individual-it — CHANGES_REQUESTED** [review]

File: config.js [hunk 7]
```diff
@@ -0,0 +1,5 @@
+exports.config = {
+    hostUrl:process.BASE_URL ?? 'http://localhost:8000',
+    adminUser:process.ADMIN_USER ?? 'admin',
+    adminPassword:process.ADMIN_PASSWORD ?? 'admin'
```
Comment (saw-jan) [inline config.js:4]: use auto-formatter

File: cucumber.conf.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  BeforeAll,
+  AfterAll,
+  After,
+  setDefaultTimeout,
+} = require("@cucumber/cucumber");
+const { chromium } = require("playwright");
+const { config } = require("./config");
+const {
+  deleteFolder,
+} = require("./tests/cucumber/acceptance/helpers/HttpHelpers");
```
Comment (saw-jan) [inline cucumber.conf.js:12]: Just asking if you want to have `tests/cucumber/acceptance` structure

File: cucumber.conf.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+setDefaultTimeout(60000);
+BeforeAll(async function () {
+  global.browser = await chromium.launch({
+    headless: config.HEADLESS,
+    slowMo: config.SLOW_MO,
+  });
+});
+
+AfterAll(async function () {
+  await deleteFolder();
```
Comment (saw-jan) [inline cucumber.conf.js:23]: cleanup in `AfterAll`? or do you want to clean up folders after each scenario?
```suggestion
AfterAll(async function () {
  await deleteFolder();
```

File: config.js [hunk 10]
```diff
@@ -0,0 +1,9 @@
+exports.config = {
```
Comment (saw-jan) [inline config.js:1]: this should go into `tests/cucumber/acceptance`. There can be other config files and there can be confusion

File: package.json [hunk 11]
```diff
@@ -0,0 +1,15 @@
+{
+  "name": "moodle-repository_ocis",
+  "version": "1.0.0",
+  "devDependencies": {
+    "@cucumber/cucumber": "^7.3.1",
+    "@cucumber/pretty-formatter": "^1.0.0",
+    "@playwright/test": "^1.39.0",
+    "axios": "^1.6.0",
+    "playwright": "^1.39.0",
+    "prettier": "^3.0.3"
```
Comment (saw-jan) [inline package.json:10]: or you need it?
```suggestion
```

File: package.json [hunk 12]
```diff
@@ -0,0 +1,15 @@
+{
+  "name": "moodle-repository_ocis",
```
Comment (saw-jan) [inline package.json:2]: how about moving all root level test related files to `tests/cucumber/acceptance`? Just my opinion

File: tests/cucumber/acceptance/features/listfolder.feature [hunk 13]
```diff
@@ -0,0 +1,18 @@
+Feature: list the resources
+  As a user
+  I want to list the resources
+  So that I can upload them to moodle
+
+  Scenario: See the list
```
Comment (saw-jan) [inline tests/cucumber/acceptance/features/listfolder.feature:6]: can be little bit more descriptive
```suggestion
  Scenario: See the list
```

File: tests/cucumber/acceptance/helpers/HttpHelpers.js [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+
+function getAdminAuthHeader() {
+  return {
+    Authorization: "Basic " + Buffer.from("admin:admin").toString("base64"),
+  };
+}
+
+async function createFolder(folderName) {
+  fileList.push(folderName);
+  return await fetch({
```
Comment (saw-jan) [inline tests/cucumber/acceptance/helpers/HttpHelpers.js:18]:
```suggestion
  return fetch({
```

File: cucumber.conf.js (hunk 8 above)
Comment (amrita-shrestha) [reply cucumber.conf.js:12]: IMO ./tests/acceptance/helpers/HttpHelpers

File: tests/cucumber/acceptance/helpers/HttpHelpers.js [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+    headers: getAdminAuthHeader(),
+  });
+}
+
+async function deleteFolder() {
+  for (value of fileList) {
+    await fetch.delete(`/dav/files/admin/${value}`, {
+      headers: getAdminAuthHeader(),
+    });
+  }
```
Comment (amrita-shrestha) [inline tests/cucumber/acceptance/helpers/HttpHelpers.js:30]: IMO we have to verify that api request is successful for create and delete folder

File: tests/cucumber/acceptance/pageObjects/Login.js [hunk 16]
```diff
@@ -0,0 +1,21 @@
+const { config } = require("../../../../config");
+class Login {
+  constructor() {
+    this.loginPageUrl = config.hostUrl + "/login/index.php";
+    this.homePageUrl = config.hostUrl + "/my/";
+    this.username = config.adminUser;
+    this.password = config.adminPassword;
+    this.usernameSelector = "#username";
+    this.passwordSelector = "#password";
+    this.loginButton = "#loginbtn";
```
Comment (amrita-shrestha) [inline tests/cucumber/acceptance/pageObjects/Login.js:10]: if we follow above pattern then
```suggestion
    this.loginButtonSelector = "#loginbtn";
```

File: tests/cucumber/acceptance/pageObjects/MyCourse.js [hunk 17]
```diff
@@ -0,0 +1,17 @@
+const { config } = require("../../../../config");
+class MyCourse {
```
Comment (amrita-shrestha) [inline tests/cucumber/acceptance/pageObjects/MyCourse.js:2]: will there be any pblm with 
```suggestion
export class MyCourse {
```

File: .gitignore [hunk 18]
```diff
@@ -1,2 +1,3 @@
-/vendor/
 composer.lock
```
Comment (amrita-shrestha) [inline .gitignore:1]: why does vendor has been removed?

**Comment by amrita-shrestha** [comment]
@S-Panta @individual-it i think we can close this PR now

**Pull request closed without merging**